File: FLEX.Data.Oracle/Oracle/OracleQueryExecutor.vb
Imports System.Data.Common
Imports System.Globalization
Imports Oracle.ManagedDataAccess.Client

Namespace Oracle
   Public NotInheritable Class OracleQueryExecutor
      Inherits QueryExecutorBase

      Private Shared ReadOnly DbFactory As DbProviderFactory = New OracleClientFactory()

      Public Overrides Sub ElaborateConnectionString(ByRef connectionString As String)
         Dim lowerConnString = connectionString.ToLower(CultureInfo.InvariantCulture)

         'Add a semicolor in order to avoid issues
         connectionString = connectionString.TrimEnd()
         If Not connectionString.EndsWith(";"c) Then
            connectionString &= ";"c
         End If

         'Connection Pooling
         If Not lowerConnString.Contains("pooling") Then
            connectionString &= "Pooling=true;"
         End If

         'Statement Cache
         If Not lowerConnString.Contains("statement cache size") Then
            connectionString &= String.Format("Statement Cache Size={0};", Configuration.Instance.OracleStatementCacheSize)
         End If
      End Sub

      Public Overrides Function OpenConnection() As IDbConnection
         Dim connection = DbFactory.CreateConnection()
         connection.ConnectionString = Common.Configuration.Instance.ConnectionString
         connection.Open()
         Return connection
      End Function
   End Class
End Namespace

File: Caravan.DataAccess/Oracle/OracleQueryExecutor.cs
using System.Data.Common;
using System.Globalization;
using Finsa.Caravan.DataAccess.Core;
using Oracle.ManagedDataAccess.Client;

namespace Finsa.Caravan.DataAccess.Oracle
{
   public sealed class OracleQueryExecutor : QueryExecutorBase
   {
      private static readonly OracleClientFactory DbFactory = new OracleClientFactory();

      public override DatabaseVendor Vendor
      {
         get { return DatabaseVendor.Oracle; }
      }

      public override void ElaborateConnectionString(ref string connectionString)
      {
         var lowerConnString = connectionString.ToLower(CultureInfo.InvariantCulture);

         // Add a semicolor in order to avoid issues
         connectionString = connectionString.TrimEnd();
         if (!connectionString.EndsWith(";"))
         {
            connectionString += ';';
         }

         // Connection Pooling
         if (!lowerConnString.Contains("pooling"))
         {
            connectionString += "Pooling=true;";
         }

         // Statement Cache
         if (!lowerConnString.Contains("statement cache size"))
         {
            connectionString += string.Format("Statement Cache Size={0};", Configuration.Instance.OracleStatementCacheSize);
         }
      }

      public override DbConnection OpenConnection()
      {
         var connection = DbFactory.CreateConnection();
         connection.ConnectionString = Configuration.Instance.ConnectionString;
         connection.Open();
         return connection;
      }
   }
}